fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -1504,4 +1597,20 @@ private static void assertConstituentPlansMatch(SyntheticRecordPlanner planner,
             assertThat(plan, matcher);
         }
     }
+
+    private static class FailIndexMaintainer extends TerribleIndexMaintainer{
```

Comment:
That feels like an extremely heavy-weight solution to the problem of wanting to quickly fail.

Just as a quick attempt, I ended up writing three classes and several hundred lines of code to get it into a ServiceLoader, whereas this particular change is fairly straightforward and obvious. 

Perhaps a cleaner approach would be to push this mocking logic into some kind of utility class, so that it's not sitting in the test code itself?

Reply:
Hm, it's maybe not the lightest weight solution I can imagine, but I'm surprised you needed three classes... It would seem that you'd need:

1. An `IndexMaintainer` implementation. For this purpose, sub-classing `ValueIndexMaintainer` and overriding just `updateIdexKeys` would seem to be sufficient
2. An `IndexMaintainerFactory` implementation for the new maintainer. The only tricky thing about that one is it needs to be marked as `@AutoService(IndexMaintainerFactory.class)`.

With those two things, the index type should now be picked up by the registry. Not counting imports and whitespace, I wouldn't expect it to exceed...100 lines? About the same order of magnitude as what is here, and more in line with the other tests that do this kind of thing

Reply:
I mispoke--I had modified three files (the test META-INF file and the other two). And a lot of lines were documentation to explain how it works and why, because it felt very disconnected from the one test it was going to be used in, so I felt it needed additional explanation.

But also, for one test it just seems like a lot to be adding in the service loader pattern here

Reply:
I see. Well, using a mocked out registry like this isn't the worst, but it does give me a bit of pause, mainly for code uniformity with the other times we wanted a non-standard index maintainer.

In terms of polluting the service loader, we actually do that fairly often in tests. For example, there are a fair number of `FunctionKeyExpression`s that get added to the service loader for one or two tests.

Reply:
I rewrote the test to force-mark the index as readable